### PR TITLE
docs: add "How to write gRPC SDK for GreptimeDB"

### PR DESCRIPTION
**Pull request by MichaelScofield**

### CODE REVIEW

**Comment by xtang** [comment]
@Wenjie0329 Please help to take a look

**Review by fengjiachun — APPROVED** [review]
LGTM

File: docs/developer-guide/how-to-write-sdk.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# How to write a gRPC SDK for GreptimeDB
+
+GreptimeDB build its own gRPC interface on top of [Apache Arrow Flight](https://arrow.apache.org/docs/format/Flight.html). If you are curious about how to write a gRPC SDK for GreptimeDB in your favorite programming language, read on!
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:3]:
```suggestion
GreptimeDB builds its gRPC interface on top of [Apache Arrow Flight](https://arrow.apache.org/docs/format/Flight.html). If you want to write a gRPC SDK for GreptimeDB in a programming language you are familiar with, read on!
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+> Currently, we have only gRPC SDK written in Java, you can find more details [here](../user-guide/java-sdk.md).
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:5]:
```suggestion
> Currently, we only have gRPC SDK written in Java, and you can find more details [here](../user-guide/java-sdk.md).
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+First you can find our protobuf definitions for GreptimeDB requests and responses in this [repo](https://github.com/GreptimeTeam/greptime-proto#for-sdk-developers). There's a section called "For SDK developers" in that repo's readme worth reading.
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:7]:
```suggestion
First, you can find our protobuf definitions for GreptimeDB requests and responses in this [repo](https://github.com/GreptimeTeam/greptime-proto#for-sdk-developers). The section named "For SDK developers" in the REAME of that repo is worth reading. 
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+GreptimeDB build its own gRPC interface on top of [Apache Arrow Flight](https://arrow.apache.org/docs/format/Flight.html). If you are curious about how to write a gRPC SDK for GreptimeDB in your favorite programming language, read on!
+
+> Currently, we have only gRPC SDK written in Java, you can find more details [here](../user-guide/java-sdk.md).
+
+First you can find our protobuf definitions for GreptimeDB requests and responses in this [repo](https://github.com/GreptimeTeam/greptime-proto#for-sdk-developers). There's a section called "For SDK developers" in that repo's readme worth reading. 
+
+Then check the programming language you are using, to see whether Arrow Flight RPC officially supports it. Currently, it supports C++, Java, Go, C# and Rust. It might add more language supports in the meantime, so stay tuned with its [Implementation Status](https://arrow.apache.org/docs/status.html#flight-rpc). If you can't find the language you are using on it, you have to write a client from sketch (starting from Arrow Flight's raw gRPC service protobuf [definition](https://arrow.apache.org/docs/format/Flight.html#protocol-buffer-definitions)).
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:9]:
```suggestion
Then check your programming language to see whether Arrow Flight RPC officially supports it. Currently, it supports C++, Java, Go, C# and Rust and might add more supported languages, so stay tuned with its [Implementation Status](https://arrow.apache.org/docs/status.html#flight-rpc). If you can't find the language you are using, you have to write a client from sketch (starting from Arrow Flight's raw gRPC service protobuf [definition](https://arrow.apache.org/docs/format/Flight.html#protocol-buffer-definitions)).
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+Now focus on the `DoGet` method of Arrow Flight gRPC service, currently all GreptimeDB requests are handled in it.
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:11]:
```suggestion
Now focus on the `DoGet` method of Arrow Flight gRPC service, which all GreptimeDB requests are handled in it.
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+The `DoGet` method is defined like this:
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:13]:
```suggestion
The `DoGet` method is defined as follow:
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+```protobuf
+  rpc DoGet(Ticket) returns (stream FlightData) {}
+```
+
+For sending a GreptimeDB request, just encode it into raw bytes, then wrap it into a `Ticket`. `Ticket` is simply a message carries bytes:
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:19]:
```suggestion
To send a GreptimeDB request, encode it into raw bytes, then wrap it into a `Ticket` which act as a message carrying bytes:
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+```protobuf
+message Ticket {
+  bytes ticket = 1;
+}
+```
+
+It's a little complicated to handle GreptimeDB responses. `DoGet` returns a stream of `FlightData`. The definition of `FlightData` is:
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:27]:
```suggestion
Handling GreptimeDB responses is a little complicated. `DoGet` returns a stream of `FlightData` and the definition of `FlightData` is:
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+   * The actual batch of Arrow data. Preferably handled with minimal-copies
+   * coming last in the definition to help with sidecar patterns (it is
+   * expected that some implementations will fetch this field off the wire
+   * with specialized code to avoid extra memory copies).
+   */
+  bytes data_body = 1000;
+}
+```
+
+Specifically the fields are:
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:58]:
```suggestion
The fields are specified as follow:
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+- `flight_descriptor` can be ignored here.
+- `data_header` must be first deserialized to `Message` (see its definition [here](https://github.com/apache/arrow/blob/master/format/Message.fbs#L134)) using [FlatBuffer](https://github.com/google/flatbuffers). The `Message`'s header type decides how the following two fields are interpreted.
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:61]:
```suggestion
- `data_header` must be first deserialized to `Message` (see its definition [here](https://github.com/apache/arrow/blob/master/format/Message.fbs#L134)) using [FlatBuffer](https://github.com/google/flatbuffers). The `Message`'s header type determines how the following two fields are interpreted.
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+- `app_metadata` carries GreptimeDB's custom data. Now this field is not empty only after client issues `InsertRequest` or `Insert Into` SQL. When it's not empty, `Message`'s header type is set to `None`. You should deserialize `app_metadata` to `FlightMetadata`. `FlightMetadata` is defined [here](https://github.com/GreptimeTeam/greptime-proto/blob/966161508646f575801bcf05f47ed283ec231d68/proto/greptime/v1/database.proto#L50).
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:62]:
```suggestion
- `app_metadata` carries GreptimeDB's custom data and this field is not empty only after client issues `InsertRequest` or `Insert Into` SQL. When it's not empty, `Message`'s header type is set to `None`. You should deserialize `app_metadata` to `FlightMetadata`. `FlightMetadata` is defined [here](https://github.com/GreptimeTeam/greptime-proto/blob/966161508646f575801bcf05f47ed283ec231d68/proto/greptime/v1/database.proto#L50).
```
Not sure the processing logic here, so didn't modified much.

File: docs/developer-guide/how-to-write-sdk.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+  - Currently `FlightMetadata` only carries "Affected Rows" when writing data into GreptimeDB, just like what "Insert Into" SQL returns in MySQL. If you don't care about the affected rows result, you can omit this `app_metadata` field. (Still, the response of `DoGet` itself should be handled as well.)
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:63]:
```suggestion
 - `FlightMetadata` only carries "Affected Rows" when writing data into GreptimeDB, just like what "Insert Into" SQL returns in MySQL. If you don't care about the result of the affected rows, you can omit the `app_metadata` field. (Still, the response of `DoGet` itself should be handled as well.)
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+- When `Message`'s header type is `Schema` or `Recordbatch`, `data_body` carries the actual data of part of the query result. You should parse all the `FlightData`s in the response stream of `DoGet` to get the complete query result. Normally, the first `FlightData` in a stream is schema, and the rests are recordbatch. You should save the first schema for later parsing the follow-up `FlightData`s.
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:64]:
```suggestion
- When `Message`'s header type is `Schema` or `Recordbatch`, `data_body` carries the actual data of part of the query result. You should parse all the `FlightData`s in the response stream of `DoGet` to get the complete query result. Normally, the first `FlightData` in a stream is schema, and the rests are recordbatch. You should save the first schema for parsing the follow-up `FlightData`s later.
```

File: docs/developer-guide/how-to-write-sdk.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        flight_data.data_body, 
+        context.schema
+      );
+      
+      result.recordbatches.push(recordbatch);
+  }
+}
+```
+
+You can also refer to our [Rust client](https://github.com/GreptimeTeam/greptimedb/blob/develop/src/common/grpc/src/flight.rs#L85).
```
Comment (Wenjie0329) [inline docs/developer-guide/how-to-write-sdk.md:112]: refer to our rust client for what? for more information or instruction?

Comment (MichaelScofield) [reply docs/developer-guide/how-to-write-sdk.md:112]: Currently we don't have an independent rust client, so I just refer to the codes.

**Comment by MichaelScofield** [comment]
@Wenjie0329 PTAL

**Review by xtang — APPROVED** [review]

**Pull request merged**